### ISSUE TITLE
[QST] can the category be on the user instead of item

### ISSUE THREAD

**Issue by lightsailpro** [NVIDIA-Merlin/Transformers4Rec#426]
# ❓ Questions & Help

## Details
If a user belongs to a site, can the site name be a category with the model favors the items that only found in all user sessions that belong to that site? I.e., if the session1 from user1 belong to site1, the top k predicted items will be from item list that is in all items filterted by site1. Number of Sites  can be in tens of thousands, so train indivisual models for each site will not be feasible. A single model that support multiple sites will be great. Item ids are actually global with each site only has a subset of item ids. Thanks

**Comment by sararb** [comment]
Thank you for your good question!   Just to confirm if I correctly understand your use case: 
1. Does it mean you have an additional context variable that will remain the same for all items within the same session? 
2. Or this variable can vary between items of the same session but when predicting the next item to recommend, you want to constrain the model to predict only from a given category? (in other terms, the model has access to additional information about the target item, which is the site name)?

**Comment by sararb** [comment]
Thank you for the clarification, This is an interesting problem! We, unfortunately, don’t have a pre-built solution for it, but here are some possible solutions I could think of: 

- Regarding your point `model handles it during the training process by filtering the site in both training and prediction.`: I think the most effective way to accomplish this would be to have a custom loss function mask predictions that do not fall within the given *site*, so that the model's parameters can be updated using only predictions from the given `site`. The evaluation step requires a post-filtering step in which you start by retrieving top-N scoring items candidates, filter the candidates of a given site and then take the final top-k (k << N) list of recommendations to present to the user. 

- For the current implementation, context-related features are fed as inputs. These features will be repeated for each item interaction. Technically, the model should take this information into account but it will require running some tests to confirm whether the model's prediction outputs are all from the same `site` or not. 

- The [Latent Cross technique](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/46488.pdf) also proved to be effective for restricting the model to predict based on a given category of the target (we used it in our [solution](https://arxiv.org/pdf/2107.05124.pdf) for the SIGIR'21 challenge). If we take your example, the idea would be to get a representation of the domain vector and merge it with the hidden representation given by the Transformer block just before applying the output layer and obtaining the logits. The context-aware block is currently being developed and will be soon included in the library.

If that makes sense to you, could you please open a feature request explaining the problem and the possible solution(s) so we can track it?

**Comment by lightsailpro** [comment]
sararb, thanks for your quick response and suggestion. Just submitted the [new feature request](https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/429). Let me know if your team need anything from me.

**Comment by gabrielspmoreira** [comment]
@lightsailpro  Just adding to this discussion, I think if the items for different sites are disjoint, then there would be no advantage of training a single model because the item embeddings are trained based on session co-occurrence and will probably end up being clustered in very different positions of the vector space for different sites. So training data from many sites altogether would not benefit accuracy, but rather introduce more complex patterns / noise for the model to learn from. And the full item catalog could lead to an very large embedding table (e.g. more than 10M embeddings) that eventually would not fit into a single GPU memory and requiring distributing the embedding table into multiple GPUs, which currently Transformers4Rec does not support.

If the objective here is only to have a very large single model to reduce the devops complexity of deploying and managing multiple small models, then I think you could take a closer look in Triton Inference Server and how you can deploy and serve multiple models in the same serving instance, if that makes sense.
